server/incoming/request.go

```diff
@@ -4,185 +4,236 @@ import (
-	pluginID  string
-
-	appID                 apps.AppID
-	actingUserID          string
+type Request struct {
+	actingUserID   string
+	destination    apps.AppID
+	sessionID      string
+	sourceAppID    apps.AppID
+	sourcePluginID string
+
+	ctx                   context.Context
+	mm                    *pluginapi.Client
+	config                config.Service
+	Log                   utils.Logger
+	sessionService        SessionService
+	requestID             string
 	actingUserAccessToken string
+	actingUser            *model.User
 }
 
-type RequestOption func(*Request)
-
-func WithCtx(ctx context.Context) RequestOption {
-	return func(r *Request) {
-		if ctx == nil {
-			panic("nil context")
-		}
-
-		r.ctx = ctx
+func NewRequest(config config.Service, log utils.Logger, session SessionService) *Request {
+	// TODO <>/<>: is the incoming Mattermost request ID available, and should it be used?
+	requestID := model.NewId()
+	return &Request{
+		ctx:            context.Background(),
+		mm:             config.MattermostAPI(),
+		config:         config,
+		Log:            log.With("request_id", requestID),
+		sessionService: session,
+		requestID:      requestID,
 	}
 }
 
-func WithTimeout(timeout time.Duration, cancelFunc *context.CancelFunc) RequestOption {
-	return func(r *Request) {
-		if timeout == 0 {
-			return
-		}
-		if cancelFunc == nil {
-			panic("nil cancel function pointer")
-		}
-
-		r.ctx, *cancelFunc = context.WithTimeout(r.Ctx(), timeout)
-	}
+// Clone creates a shallow copy of request, allowing clones to apply per-request changes.
+func (r *Request) Clone() *Request {
+	clone := *r
+	return &clone
 }
 
-func WithAppContext(cc apps.Context) RequestOption {
-	return func(r *Request) {
-		if cc.ActingUser != nil {
-			r.SetActingUserID(cc.ActingUser.Id)
-		}
-		r.actingUserAccessToken = cc.ActingUserAccessToken
-	}
+func (r *Request) Ctx() context.Context {
+	return r.ctx
 }
 
-func WithAppID(appID apps.AppID) RequestOption {
-	return func(r *Request) {
-		r.SetAppID(appID)
-	}
+func (r *Request) WithCtx(ctx context.Context) *Request {
+	r = r.Clone()
+	r.ctx = ctx
+	return r
 }
 
-func NewRequest(mm *pluginapi.Client, config config.Service, log utils.Logger, session SessionService, opts ...RequestOption) *Request {
-	r := &Request{
-		ctx:            context.Background(),
-		mm:             mm,
-		config:         config,
-		Log:            log,
-		sessionService: session,
-
-		// TODO <>/<>: is the incoming Mattermost request ID available, and should it be used?
-		requestID: model.NewId(),
+func (r *Request) WithTimeout(timeout time.Duration, cancelFunc *context.CancelFunc) *Request {
+	if timeout == 0 {
+		return r
 	}
-
-	r.Log = r.Log.With(
-		"request_id", r.requestID,
-	)
-
-	for _, opt := range opts {
-		opt(r)
+	r = r.Clone()
+	ctx, f := context.WithTimeout(r.Ctx(), timeout)
+	r.ctx = ctx
+	if cancelFunc != nil {
+		*cancelFunc = f
 	}
+	return r
+}
 
+func (r *Request) WithDestination(appID apps.AppID) *Request {
```

Comment:
Wonder if we'll have a destination plugin_id in the near future, similar to the source_plugin_id..

Reply:
It is already the case, but the requests are still routed to it as an app, and then the app's DeployType `plugin` instructs to call the plugin "directly" (fake HTTP over RPC, through the server). The (destination) plugin ID is only used in the `upstream` and needs not be a part of the `Request`. It's "just an app".